src/visitable.cpp

```diff
@@ -72,21 +72,30 @@ bool visitable<T>::has_item_with( const std::function<bool( const item & )> &fil
     } ) == VisitResponse::ABORT;
 }
 
+// Sums the two terms, being careful to not trigger any overflows.
+// Assumes that all terms are positive and base <= limit.
+static long accumulate_with_limit( long base, long addend, long limit ) {
+    if( addend > limit - base ) {
+        return limit;
+    }
+    return base + addend;
+}
+
 template <typename T>
 static int has_quality_internal( const T& self, const quality_id &qual, int level, int limit )
 {
     int qty = 0;
 
     self.visit_items( [&qual, level, &limit, &qty]( const item *e ) {
         if( e->get_quality( qual ) >= level ) {
-            qty += e->count_by_charges() ? e->charges : 1;
+            qty = accumulate_with_limit( qty, e->count_by_charges() ? e->charges : 1, limit );
```

Comment:
You're confusing "verbose" with "obfuscated", the helper is completely explicit about what it's doing, which is precisely why I don't like using an overloaded operator for it, THAT would be obfuscated, since there would be no sign that it's happening in the expression itself.
I'm not particularly excited about re-hacking the hack by tweaking the value, since I haven't the slightest idea what side effects that would have, whereas with this change I'm certain that I'm simply correcting the code to have the originally intended behavior.